app/components/generic-chart.js

```diff
@@ -0,0 +1,293 @@
+    tPubChanged: Ember.observer('tPub', function() {
+        let otherSubsets = (this.get('tPre') || this.get('tCre') || this.get('tPro'));
+        if(this.get('tPub')) { // if the user checked the box
+            this.filterTS('publication',otherSubsets);
+        }
+        else { // if the user unchecked the box
+            if(!otherSubsets) {
+                let data = this.get('timeseriesList');
+                this.updateChart(data);
+            }
+            else {
+                this.popTS('publication');
+            }
+        }
+    }),
+
+    // If the user wants to isolate creativeworks:
+    tCreChanged: Ember.observer('tCre', function() {
+        let otherSubsets = (this.get('tPre') || this.get('tPub') || this.get('tPro'));
+        if(this.get('tCre')) { // if the user checked the box
+            this.filterTS('creativework',otherSubsets);
+        }
+        else { // if the user unchecked the box
+            if(!otherSubsets) {
+                let data = this.get('timeseriesList');
+                this.updateChart(data);
+            }
+            else {
+                this.popTS('creativework');
+            }
+        }
+    }),
+
+    // If the user wants to isolate projects:
+    tProChanged: Ember.observer('tPro', function() {
+        let otherSubsets = (this.get('tPre') || this.get('tPub') || this.get('tCre'));
+        if(this.get('tPro')) { // if the user checked the box
+            this.filterTS('project',otherSubsets);
+        }
+        else { // if the user unchecked the box
+            if(!otherSubsets) {
+                let data = this.get('timeseriesList');
+                this.updateChart(data);
+            }
+            else {
+                this.popTS('project');
+            }
+        }
+    }),
+
+    // Isolate specific subsets of abstractcreativeworks
+    filterTS(typeString,o) {
+        let d = this.get('aggregations.articles_over_time.buckets');
+        let firstRow = ['x'];
+        let secondRow = [typeString];
+        d.forEach(function(entry) {
+            firstRow.push(entry.key_as_string);
+            let hasPubBucket = false;
+            entry.arttype.buckets.forEach(function(b) {
+               if(b.key === typeString) {
+                   hasPubBucket = true;
+                   secondRow.push(b.doc_count);
+               }
+            });
+            if(!hasPubBucket) {
+               secondRow.push(0);
+            }
+        });
+        if(o) {
+            this.pushTS(secondRow);
+        }
+        else {
+            let data = [firstRow, secondRow];
+            this.updateChart(data);
+        }
+    },
+
+    dataChanged: Ember.observer('aggregations', function() {
+        this.updateChart();
+    }),
+
+    data: [],
+
+    sizeChanged: Ember.observer('resizedSignal', function() {
+        this.updateChart();
+    }),
+
+    charTypeChange: Ember.observer('chartType', function(){
+        this.updateChart();
+    }),
+
+    updateChart() {
```

Comment:
lets see it theres a way to refactor this without the switch. For example `this.set('data', this.get('aggregations.contributors.buckets'));` could be `this.set('data', this.get(the_aggregation));` where `the_aggregation` is a string  defined in the settings panel of the widget.